### ISSUE TITLE
`bundle install` in a gem project may unexpectedly update gem dependencies after changing your gem's version

### ISSUE THREAD

**Issue by myronmarston** [rubygems/rubygems#6182]
### Describe the problem as clearly as you can

[According to the bundler docs](https://bundler.io/v2.3/man/bundle-install.1.html#CONSERVATIVE-UPDATING):

> When you make a change to the [Gemfile(5)](https://bundler.io/v2.3/man/gemfile.5.html) and then run `bundle install`, Bundler will update only the gems that you modified.

...but I've found a case where this doesn't hold true.  After changing the version number of my gem, when I run `bundle install`, it updates a dependency of my gem even though I didn't make any change related to that.

### Did you try upgrading rubygems & bundler?

Yes. I can reproduce this bug on the following versions:

```
$ gem -v
3.3.26
$ bundler -v
Bundler version 2.3.26
```

### Post steps to reproduce the problem

These reproduction steps use a [gist](https://gist.github.com/myronmarston/23d8a1bf7140820cb5205127b84b1a09) I setup with an example.

```bash
$ git clone https://gist.github.com/myronmarston/23d8a1bf7140820cb5205127b84b1a09 bundler_bug_repro
$ cd bundler_bug_repro
$ bundle install
```

At this point you'll have a bundle with version `2.0.15` of the `graphql` gem.

Make the following edit locally:

```diff
diff --git a/bundler_bug_repro.gemspec b/bundler_bug_repro.gemspec
index bc641db..7e47e60 100644
--- a/bundler_bug_repro.gemspec
+++ b/bundler_bug_repro.gemspec
@@ -2,7 +2,7 @@
 
 Gem::Specification.new do |spec|
   spec.name = "bundler_bug_repro"
-  spec.version = "0.1.0"
+  spec.version = "0.1.1"
   spec.authors = ["Someone"]
   spec.email = ["someone@gmail.com"]
   spec.summary = "Reproducing a bundler bug."
```

Then run:

```bash
$ bundle install
```

This produces the following output:

```
Resolving dependencies...Fetching gem metadata from https://rubygems.org/............
...
Using bundler 2.3.26
Fetching graphql 2.0.16 (was 2.0.15)
Installing graphql 2.0.16 (was 2.0.15)
Using bundler_bug_repro 0.1.1 (was 0.1.0) from source at `.`
Bundle complete! 1 Gemfile dependency, 3 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
```

Notably, it upgraded the `graphql` gem from `2.0.15` to `2.0.16` even though I did not ask it to or make any change related to that.  In this case, I had just bumped the version of my gem and was running `bundle install` to get that updated version recorded in `Gemfile.lock`. I do not expect that it'll also update some of my dependencies.

### Which command did you run?

`bundle install`

### What were you expecting to happen?

I expected it to not update any gems, and to change my `Gemfile.lock` like so:

```diff
diff --git a/Gemfile.lock b/Gemfile.lock
index a113c4f..af62376 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -1,13 +1,13 @@
 PATH
   remote: .
   specs:
-    bundler_bug_repro (0.1.0)
+    bundler_bug_repro (0.1.1)
       graphql (~> 2.0)

 GEM
```

### What actually happened?

It updated `graphql` to `2.0.16` and changed my `Gemfile.lock` like so:

```diff
diff --git a/Gemfile.lock b/Gemfile.lock
index a113c4f..af62376 100644
--- a/Gemfile.lock
+++ b/Gemfile.lock
@@ -1,13 +1,13 @@
 PATH
   remote: .
   specs:
-    bundler_bug_repro (0.1.0)
+    bundler_bug_repro (0.1.1)
       graphql (~> 2.0)

 GEM
   remote: https://rubygems.org/
   specs:
-    graphql (2.0.15)
+    graphql (2.0.16)

 PLATFORMS
   ruby
```

## Environment

```
Bundler             2.3.26
  Platforms         ruby, arm64-darwin-21
Ruby                2.7.5p203 (2021-11-24 revision f69aeb83146be640995753667fdd6c6f157527f5) [arm64-darwin-21]
  Full Path         /Users/myron/.rvm/rubies/ruby-2.7.5/bin/ruby
  Config Dir        /Users/myron/.rvm/rubies/ruby-2.7.5/etc
RubyGems            3.3.26
  Gem Home          /Users/myron/.rvm/gems/ruby-2.7.5
  Gem Path          /Users/myron/.rvm/gems/ruby-2.7.5:/Users/myron/.rvm/rubies/ruby-2.7.5/lib/ruby/gems/2.7.0
  User Home         /Users/myron
  User Path         /Users/myron/.gem/ruby/2.7.0
  Bin Dir           /Users/myron/.rvm/gems/ruby-2.7.5/bin
Tools               
  Git               2.37.0
  RVM               1.29.12 (latest)
  rbenv             not installed
  chruby            not installed
  rubygems-bundler  (1.4.5)
```

## Bundler Build Metadata

```
Built At          2022-11-17
Git SHA           23ec5b8501
Released Version  true
```

## Bundler settings

```
build.eventmachine
  Set for the current user (/Users/myron/.bundle/config): "--with-opt-dir=/opt/homebrew/opt/openssl"
build.puma
  Set for the current user (/Users/myron/.bundle/config): "--with-opt-dir=/opt/homebrew/opt/openssl"
gem.changelog
  Set for the current user (/Users/myron/.bundle/config): false
gem.ci
  Set for the current user (/Users/myron/.bundle/config): false
gem.coc
  Set for the current user (/Users/myron/.bundle/config): false
gem.linter
  Set for the current user (/Users/myron/.bundle/config): false
gem.mit
  Set for the current user (/Users/myron/.bundle/config): false
gem.push_key
  Set for the current user (/Users/myron/.bundle/config): "[REDACTED]"
gem.test
  Set for the current user (/Users/myron/.bundle/config): "rspec"
```

### Gemfile

```ruby
# frozen_string_literal: true

source "https://rubygems.org"

# Specify your gem's dependencies in bundler_bug_repro.gemspec
gemspec
```

### Gemfile.lock

```
PATH
  remote: .
  specs:
    bundler_bug_repro (0.1.0)
      graphql (~> 2.0)

GEM
  remote: https://rubygems.org/
  specs:
    graphql (2.0.15)

PLATFORMS
  ruby

DEPENDENCIES
  bundler_bug_repro!

BUNDLED WITH
   2.3.26
```

### bundler_bug_repro.gemspec

```ruby
# frozen_string_literal: true

Gem::Specification.new do |spec|
  spec.name = "bundler_bug_repro"
  spec.version = "0.1.1"
  spec.authors = ["Someone"]
  spec.email = ["someone@gmail.com"]
  spec.summary = "Reproducing a bundler bug."

  spec.required_ruby_version = ">= 2.6.0"

  # Specify which files should be added to the gem when it is released.
  # The `git ls-files -z` loads the files in the RubyGem that have been added into git.
  spec.files = Dir.chdir(File.expand_path(__dir__)) do
    `git ls-files -z`.split("\x0").reject do |f|
      (f == __FILE__) || f.match(%r{\A(?:(?:test|spec|features)/|\.(?:git|travis|circleci)|appveyor)})
    end
  end

  spec.bindir = "exe"
  spec.executables = spec.files.grep(%r{\Aexe/}) { |f| File.basename(f) }
  spec.require_paths = ["lib"]

  spec.add_dependency "graphql", "~> 2.0"
end
```

**Comment by deivid-rodriguez** [comment]
Thanks for the report @myronmarston!

I had observed this myself in the past but never got around to investigating it. Now I have a nice repro, so I will :)

**Comment by deivid-rodriguez** [comment]
@myronmarston The fix has been released with Bundler 2.4!

**Comment by myronmarston** [comment]
Thanks for tackling this and cutting a release with the fix so fast!
